File: Codeforces/Camp Contest 2/B_Divisibility_by_2_n.py
from functools import reduce, cache

t = int(input())

for _ in range(t):
    n = int(input())
    a = list(map(int, input().split()))
    MOD = 2 ** n
    prod = reduce(lambda x, y: x * y, a)
    if prod % MOD == 0:
        print(0)
        continue

    @cache
    def count_twos(x):
        res = 0
        while x % 2 == 0:
            res += 1
            x //= 2
        return res

    twos = count_twos(prod)
    to_add_twos = [count_twos(x + 1) for x in range(1, n, 2)]
    av = sum(to_add_twos)

    if twos + av < n:
        print(-1)
        continue

    target = n - twos
    dp = [float('inf')] * (target + 1)
    dp[0] = 0

    for num in to_add_twos:
        for i in range(target, num - 1, -1):
            dp[i] = min(dp[i], dp[i - num] + 1)

    if dp[target] == float('inf'):
        print(-1)
    else:
        print(dp[target])

File: Codeforces/Camp Contest 2/B_Divisibility_by_2_n.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <numeric>
#include <unordered_map>

using namespace std;

unordered_map<int, int> memo;

int count_twos(int x)
{
    // base case: if x is not divisible by 2, return 0
    if (x % 2 != 0) return 0;
    if (memo.count(x)) return memo[x];
    int res = 1 + count_twos(x / 2);
    memo[x] = res;
    return res;
}

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n;
        cin >> n;
        vector<int> a(n);
        for (int i = 0; i < n; i++)
        {
            cin >> a[i];
        }
        int MOD = pow(2, n);
        int prod = 1;
        for (int i = 0; i < n; i++)
        {
            prod *= a[i];
        }
        if (prod % MOD == 0)
        {
            cout << 0 << endl;
            continue;
        }
        int twos = count_twos(prod);
        vector<int> to_add_twos(n/2);
        for (int i = 1; i < n; i += 2)
        {
            to_add_twos.push_back(count_twos(i + 1));
        }
        int av = accumulate(to_add_twos.begin(), to_add_twos.end(), 0);
        if (twos + av < n)
        {
            cout << -1 << endl;
            continue;
        }
        int target = n - twos;
        sort(to_add_twos.rbegin(), to_add_twos.rend());
        int operations = 0;
        int total = 0;
        for (int num : to_add_twos)
        {
            if (total >= target)
            {
                break;
            }
            total += num;
            operations += 1;
        }
        if (total < target)
        {
            cout << -1 << endl;
            continue;
        }
        cout << operations << endl;
    }
    return 0;
}